Feature: New order list management + MVVM

AmandaRiu opened the pull request:
This PR completes the project to convert the order list tab over to using the new order list management functionality in FluxC and closes #1330. 

The most noticeable changes are how list loading works. The loading skeleton is now visible at the list item level so as items are fetched they are loaded instead of waiting for the entire results to be loaded. Loading is also smoother and faster. 

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/5810477/68819156-19ddeb80-063c-11ea-99e1-7a5ff005c03c.gif)

Since this is a feature branch, most functionality has already been tested and approved. The only new code in this branch is:
- Refreshing this feature branch with the `develop` branch forced me to manually move the payment gateway code added to `OrderListPresenter` to `OrderListViewModel` in 1e30ffcdc8c86380a88313677ee5877b06b295d6. So for the purpose of this PR, special attention should be made to the refunds functionality found in #1496. 
- Added a fix to refresh the active order list when a new order notification is received in f5ba50553da899d7d5943a081df0263eea26e8a2
- Updated the logic to load the list even if the `OrderListFragment` is not active in 9c3d242fed4e1a2bc7c5f1555aafb107300a5c8f. With the new list management design, there is no need to wait for the fragment to be active to start the process of loading the list.  Also removed the duplicate call to load the list while initializing the viewModel.
- Code cleanup to remove unused code in 35a857378f7a1ed579ec2d3006285786907d8ec8

## Smoke Test

- Loading orders, pull to refresh, load more
- Moving between ALL and PROCESSING tabs show the appropriate data
- Opening an order to view the detail
- List updated when an order status is changed in order detail
- Offline handling
- Searching orders
- Filtering by Order Status
- New Order Notifications:
  - If app already open and orders visible, the order list should refresh when new notification received.
  - opening a new order notification when app already open. 
  - opening a new order notification when app is closed.
- Empty states:
  - All orders processed
  - No orders yet (if store has no orders)
  - No orders to process yet (if store has no orders)
- Issuing a refund on an order


Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListItemUIType.kt:
```diff
@@ -0,0 +1,71 @@
+        val currencyCode: String
+    ) : OrderListItemUIType()
+
+    override fun equals(other: Any?): Boolean {
```
If you make every child class a `data` class, you'll get the `equals()` operation completely for free.

0nko replied:
@AmandaRiu and here :).

AmandaRiu replied:
Awesome catch!. Fixed in 7fd3d56eee5971e20db6dcf99aa44575a6cb3064

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderFetcher.kt:
```diff
@@ -0,0 +1,86 @@
+ * Once all orders have been successfully fetched, any `PagedList` views observing this specific type of
+ * data will update automatically.
+ */
+@Singleton
```
I'm not completely sure why this class needs to be a singleton. It's only used inside `OrderListItemDataSource`, which is only used inside `OrderListViewModel`. In fact, this class is not even being injected, it's manually constructed [here](https://github.com/woocommerce/woocommerce-android/pull/1559/files#diff-5d71a6b23c0503fd31eb3a9973b46598R41).

0nko replied:
@AmandaRiu Not sure if you saw this comment above. Just making sure you didn't miss it.

AmandaRiu replied:
@0nko Oh wow, I did miss this. I'll check it out now. Thank you!

AmandaRiu replied:
Great question and nice catch. As I mentioned in another comment: 

> I made `OrderFetcher`  a singleton because of the functionality behind the class. It’s responsible for keeping track of ongoing network requests to fetch orders. Having a single instance avoids issues of processing multiple requests for the same data.  This may not seem like a big deal now, but imagine in the future if some other area in the app is using this functionality, then we’d have some unexpected race conditions with orders (this future may be really soon too since in order to avoid the refresh of the view when switching between tabs I may need to have two different `PageListWrapper`s).

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListFragment.kt:
```diff
@@ -31,85 +33,110 @@ import com.woocommerce.android.ui.base.TopLevelFragment
-    override var isRefreshing: Boolean = false
-    override var isSearching: Boolean = false
+    private val viewModel: OrderListViewModel by lazy {
+        ViewModelProviders.of(requireActivity(), viewModelFactory).get(OrderListViewModel::class.java)
```
Just curious, why is the ViewModel scoped to the activity? Doesn't the fragment (being a `TopLevelFragment`) live just just as long?

AmandaRiu replied:
You make a very good point. I’ve fixed this in c25fe8fb359964c2d3945a34ee60209d67969434

JavonDavis commented:
Hey! Bumping this to `3.2`, if you'd like to get it in for this release please merge this into release/3.1 and then DM me to release a new Beta version.`

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt:
```diff
@@ -0,0 +1,344 @@
+
+        listenToEmptyViewStateLiveData(pagedListWrapper)
+
+        _pagedListData.addSource(pagedListWrapper.data) { pagedList ->
```
All these `MediatorLiveData` properties should really be [transformations](https://developer.android.com/reference/androidx/lifecycle/Transformations.html). `MediatorLiveData` is intended to have multiple triggers, whereas a transformation takes the input LiveData, makes some changes (or not) and passes them along to the original LifeCycleOwner.

AmandaRiu replied:
@0nko I'm not really sure I understand this suggestion. According to the docs for `MediatorLiveData`:

> {@link LiveData subclass which may observe other {@code LiveData} objects and react on {@code OnChanged} events from them.

which is exactly what my code is doing. I'm observing different `LiveData` events on `PagedListWrapper` and passing them along to other `LiveData` streams. For example, consider `_emptyViewState`:

```
private fun listenToEmptyViewStateLiveData(wrapper: PagedListWrapper<OrderListItemUIType>) {
    _emptyViewState.addSource(wrapper.isEmpty) { createAndPostEmptyUiState(wrapper) }
    _emptyViewState.addSource(wrapper.isFetchingFirstPage) { createAndPostEmptyUiState(wrapper) }
    _emptyViewState.addSource(wrapper.listError) { createAndPostEmptyUiState(wrapper) }
}
```

The source is different than the destination and I rely on multiple sources to feed into that final `_emptyViewState` `LiveData` object. There are 1:1 versions like with `_pagedListData`, but they are necessary because the `pagedListWrapper` instance changes every time a new list loads so the associated `LiveData` observations also change. Because of this I'm not sure how `Transformations` will fit here since it'd have to be setup before I set the public `pagedListData` stream that `OrderListFragment` will subscribe to. Using `MediatorLiveData.addSource()` in this scenario ensures the main `_pagedListData` variable stays the same and therefore this will continue to work:
```
val pagedListData: LiveData<PagedOrdersList> = _pagedListData
```

I may not fully be grasping your suggestion however so I'll be happy to revisit if you could provide some code direction here.

0nko replied:
> There are 1:1 versions like with _pagedListData, but they are necessary because the pagedListWrapper instance changes every time a new list loads so the associated LiveData observations also change.

The `_emptyViewState` with multiple sources is good, I meant those 1:1 versions. But as you've mentioned, the `Transformations` would need to be set up during the ViewModel instantiation, so these make sense too. Thanks for the explanation 👍.

jkmassel commented:
We're freezing `3.2` today, so I'm moving these forward to `3.3`. If you'd like them released as part of `3.3`, please adjust the target from `develop` to `release/3.3`, and let me know once this PR has been merged – I'll be happy to ship a new release candidate!

AmandaRiu commented:
@0nko Thank you for the review! I've addressed your code comment and I've gone in and merged the `SavedState` changes from `develop` (as well as later updated my branch with the latest and greatest from develop).

`OrderListViewModel` now uses the new `SavedState` logic for the following state-level variables:
* `arePaymentGatewaysFetched`
* `isRefreshPending`

The rest of the state-level variables are being handled by the fragment. In the future I’d like to move those over as well, but when I tried to implement the changes I started having a ton of issues with the UI so I’d like to leave that for later since it’s going to require bigger changes than I wanted to implement in this feature PR and should not prevent this feature from being merged into `develop`.

Making this change removed the `start()` method in `OrderListViewModel` so I also had to tweak the logic in `OrderListFragment` to ensure the order status options are properly downloaded and populated. 

I also fixed the instrumented and unit tests for the order list view.  Other tests for order detail, notes, etc, are failing, but they are failing in `develop` as well and are likely due to all the recent changes to move each of these views over from using a `Presenter` to using a `ViewModel`.

I squashed all these changes into commit 74601e215c25ee2972f94b61687e2d0290da4978.

I definitely recommend going through the recommended smoke testing I listed out in the PR descriptions to ensure none of these changes messed up anything. While I was in there I also fixed a localization issue not related specifically to this feature, but one that would be really annoying for anyone loading the list in a different language. You can test this by:

1. Change the device language to something other than english. 
2. Open the app and switch to the "orders" list. Click on the "Processing" tab. 
3. Close the app.
4. Reopen the app and switch to the "Orders" list. Only orders labeled as "processing" should load in the list.

I also recommend:
- Run tests in `OrderListViewModelTests`
- Run tests in `OrderListItemTest`
- Run tests in `OrderListNavigationTest`  

Ready for another review!

AmandaRiu commented on WooCommerce/src/androidTest/kotlin/com/woocommerce/android/helpers/CoroutineTestUtils.kt:
```diff
@@ -0,0 +1,32 @@
+import kotlin.coroutines.EmptyCoroutineContext
+import kotlin.coroutines.resume
+
+fun <T> test(context: CoroutineContext = EmptyCoroutineContext, block: suspend CoroutineScope.() -> T) {
```
That's for the unit tests, this one is for the instrumentation tests. 😄

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListFragment.kt:
```diff
@@ -581,79 +572,53 @@ class OrderListFragment : TopLevelFragment(), OrderListContract.View,
      */
-    override fun submitSearchDelayed(query: String) {
+    private fun submitSearchDelayed(query: String) {
         searchHandler.postDelayed({
```
@AmandaRiu, same here.

AmandaRiu replied:
I love this and definitely want to do both, but it’s not a blocker and I’d like to get this very large project merged so I’ve [created a ticket](https://github.com/woocommerce/woocommerce-android/issues/1647) for it to be handled later.

0nko commented:
Thanks for the changes, Amanda! Things look good 👍.

- I pinged you on a couple of notes without answer just to make sure you accidentally didn't miss them.
- I smoke-tested the order list and I noticed that it's being reloaded every time the tab changes, even if the contents of the list doesn't. There is also some delay between the tab change and the list change. You can see it in this short video:
<img src="https://user-images.githubusercontent.com/1522856/70051196-e0f8bd00-15d0-11ea-9bfd-03a9de9c7099.gif" width="400" />

- When I ran the `OrderListItemTest` I got 2 failures: `verifyOrderListItemOrderStatusLabelDisplayedCorrectly` and `verifyOrderListItemTotalDisplayedCorrectlyForMultipleCurrencies`:

```androidx.test.espresso.NoMatchingViewException: No views in hierarchy found matching: RecyclerView with id: com.woocommerce.android.dev:id/ordersList at position: 6```

- When I use pull-to-refresh, the refresh indicator seems to be stopped & started again:
<img src="https://user-images.githubusercontent.com/1522856/70051691-dbe83d80-15d1-11ea-8283-26bfda8da088.gif" width="400" />

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListFragment.kt:
```diff
@@ -31,85 +33,110 @@ import com.woocommerce.android.ui.base.TopLevelFragment
+    private var isSearching: Boolean
+        private set(value) { viewModel.isSearching = value }
+        get() = viewModel.isSearching
+    var isRefreshing: Boolean = false
```
I love this and definitely want to do both, but it’s not a blocker and I’d like to get this very large project merged so I’ve [created a ticket](https://github.com/woocommerce/woocommerce-android/issues/1647) for it to be handled later. 👍

AmandaRiu commented:
@0nko Thank you for such a thorough review. My apologies for missing a bunch of the items from the first round of your review. 🙈 I went ahead and addressed/fixed all those original comments. Below is a response to your initial review summary. I figured I'd pop this back over to you so you can review it while I work through the issues you found in the second round. Also note that I've updated this PR with changes in `develop` and the tests are failing due to recent changes in `MockedProductDetailViewModel` that attempt to override a final variable:

```
    override val viewStateData: LiveDataDelegate<ViewState> =
            LiveDataDelegate(arg0, ViewState(), "", onChange = {
                combineData(it.product!!, Parameters("$", "oz", "in"))
            })
```

I just comment all that out to run the order list tests and they are all successful so I'll need to do more digging to figure out why they are failing for you. 

## Round 1 (missed items)

> The thing that I’m not completely sure about is why the `OrderListViewModel` needs to be a `LifeCycleOwner` and why the `DataSource` and `OrderFetcher` explicitly follow its events. Why can’t the `ViewModel` just do the clean up itself in `onCleared()`?

Those are really great questions! I'll address them individually:
	
> why the `OrderListViewModel` needs to be a `LifeCycleOwner`

This is a requirement of the `PagedListWrapper` . That’s the `ListStore` code we are using in FluxC (shared by WPAndroid). You can see it being used to create new lists like this:

`val pagedListWrapper = listStore.getList(listDescriptor, dataSource, lifecycle)`

> why the `DataSource` and `OrderFetcher` explicitly follow its events. Why can’t the `ViewModel` just do the clean up itself in `onCleared()`?

Yeah I can see why you found this to be really confusing, it took me some time to figure it out just now because my implementation of this works, but isn’t really valid. I made `OrderFetcher`  a singleton because of the functionality behind the class. It’s responsible for keeping track of ongoing network requests to fetch orders. Having a single instance avoids issues of processing multiple requests for the same data. This may not seem like a big deal now, but imagine in the future if some other area in the app is using this functionality, then we’d have some unexpected race conditions with orders (for example, I think it would be a better experience for the user to not see the list filter jump when switching between filters and this would require multiple `PageListWrapper`s).  I had this in mind when I made it a singleton and also injectable, but lost that train of development thought somewhere between all the heavy breaks in this project. This was all muddied by the fact that I was sending in the lifecycle of the `OrderListViewModel` which is not a `Singleton`  so that wouldn’t make sense.  I’ve made this class a proper `Singleton` and removed the unnecessary lifecycle dependencies for both `OrderListItemDatasource` and `OrderFetcher` in 7d67ccc7e9977e3b6ead88ca6fbf080938fbc45a

Just a note. I was using the `lifecycle` dependency in `OrderFetcher` to unregister from the `dispatcher`, but now that this class is a proper singleton this is no longer necessary.

AmandaRiu commented:
@0nko I've finished all my fixes! Yay! 

> Order list reloaded every time the tab changes, even if the contents of the list doesn’t. There is also some delay between the tab change and the list change.

This has been bugging me for the entire project and is due to the following:
1. The tabs and the filter/search results all share the same list view.
2. Every time the `orderStatusFilter` or  `query` values to fetch a list change, a new `pagedListWrapper` needs to be generated. Since we observe the `isFetchingFirstPage` and display the `refreshing` spinner whenever it changes to `true`, we look like we are refreshing every time we change between tabs or filter/search mode. 

I’ve fixed this in 2246efe66cdb4724bc82c50e3d15233d69690e4b by creating a separate `pagedListWrapper` for the all, processing, and search/filter list types. The “all” and “processing” lists will re-use their own `pagedListWrapper` so the experience is more seamless. Note: This [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1441) must first be merged before these changes can be approved. 

>When I ran the `OrderListItemTest` I got 2 failures

I’m unable to duplicate this issue and have since made further changes to support the fixes from above and all passes fine. To avoid having this PR opened longer than necessary, if you are still seeing this failure please let me know and I’ll open a separate ticket for it, or if you like you can open the ticket and assign it to me. 

>When I use pull-to-refresh, the refresh indicator seems to be stopped & started again.

Fixed in d100219c7528628b25f21808b3466b8d84d1bc8e

That should cover all your findings. Just a heads up just in case you run into this during testing. If you were to go offline in the app and then open the “Search” screen and perform a search, you will likely get an “offline” empty state error unless you had performed that exact search before. This can happen *even if there are cached records matching your search term in the db.* This is a function of the way the `listStore` works and is not technically a bug. Since this PR has already been open for far too long and this feature branch already has a bunch of direct changes, [I’ve gone ahead and opened a separate issue](https://github.com/woocommerce/woocommerce-android/issues/1656) for it. That shouldn’t block this PR at all. For more background on this behavior issue, see `p99K0U-2b3-p2`.

Ready for another round 💣

0nko commented:
Btw, I noticed that wordpress-mobile/WordPress-FluxC-Android#1441 has already been merged, so we could remove the `Not ready for merge` label here, right?

AmandaRiu commented:
@0nko 

> I remember this was mentioned in one of the P2s but I noticed that when a new order is made, it appears under Processing (with an Upcoming header), but not under All orders. Just wanna make sure it's working as expected.

Future orders will show up in "All" orders, but not in the "Processing" section - that's the only scenario I can think of where that will happen.

0nko commented:
> Future orders will show up in "All" orders, but not in the "Processing" section - that's the only scenario I can think of where that will happen.

I read your comment again and realized the exact opposite is happening:

![image](https://user-images.githubusercontent.com/1522856/70646280-d9778a80-1c46-11ea-85ba-8b5f8cdec25a.png)

![image](https://user-images.githubusercontent.com/1522856/70646310-ee541e00-1c46-11ea-8001-cbced72eaf45.png)

AmandaRiu commented:
> I read your comment again and realized the exact opposite is happening:

I got it backwards in my explanation. Just verified the future order is showing properly.
